primitives/trie/src/storage_proof.rs

```diff
@@ -26,6 +26,7 @@ use hash_db::{Hasher, HashDB};
 /// the keys covered by the proof. Verifying the proof requires constructing the partial trie from
 /// the serialized nodes and performing the key lookups.
 #[derive(Debug, PartialEq, Eq, Clone, Encode, Decode)]
+#[cfg_attr(feature = "std", derive(serde::Serialize, serde::Deserialize, Hash, PartialOrd, Ord))]
```

Comment:
Using that type directly will end up with extremely unfriendly encoding of `Vec<u8>`. We should convert to a RPC specific type and wrap as `Bytes`, which will use hex-formatted strings for the nodes.

Reply:
I will try.

Reply:
Hi @tomusdrw , I found that `StorageKey` and `StorageData` also use `Vec<u8>`. Should `StorageKey`, `StorageData` and `StorageProof` be updated to `Bytes` together in another PR? If it shouldn't be, can I make changes on the struct `StorageProof` in sp_tire?

Reply:
As it would better now to include the block hash that you used to generate the proof, I propose that you generate a custom struct here in RPC that uses `Bytes` for everything.

Reply:
`StorageKey` and `StorageData` use a custom serialization/deserialization function, so they end up as hex:
https://github.com/paritytech/substrate/blob/57221630ed3efd182bce19468d211c0d98454d05/primitives/storage/src/lib.rs#L31

Actually you can use `StorageData` or `StorageKey`, but I don't think it makes sense semantically, right? Cause the proof contains trie nodes, not the storage data.

I second Basti on the custom struct.

Reply:
The last commit is an attempt I made.
In my scenario, the person who request the proof and the person who verified the proof are the same person. I'm not sure whether the block hash should be included in the proof.

Reply:
```
struct ReadProof<B: BlockT> {
    at: B::Hash,
    proof: StorageProof,
}
```

Reply:
> ```
> struct ReadProof<B: BlockT> {
>     at: B::Hash,
>     proof: StorageProof,
> }
> ```

Updated. I checked using curl:
```
$ curl -H "Content-Type: application/json" -d '{"id":1, "jsonrpc":"2.0", "method": "state_getReadProof", "params": [["0x5c0d1176a568c1f92944340dbfed9e9c530ebca703c85910e7164cb7d1c9e47b"]]}' http://localhost:9933/
{"jsonrpc":"2.0","result":{"at":"0xcc57830e92ca3cdb0a589fda70fb6614c4b8c2932816cfb7fb993c577aa4d122","proof":["0x7e0d1176a568c1f92944340dbfed9e9c530ebca703c85910e7164cb7d1c9e47b809ee5e5bdc0ec239eb164f865ecc345ce4c88e76ee002e0f7e318097347471809","0x803e9180def4a9c444d76421177b0cd327c0ab2e724ef32139086024fb5d18f9056693d7809fd2fab99d141ca4dcc0cefa8422a1965c59abcc8b1494fbe1922978a9a22de980d6cb3508530adfc31a1e38f3eac7feefc71514e5bffa444abd0ff9056e4493c98063c757a64330433a3eaa2810daf132978347e5f67df1dd1f7cef24f090e98b8c80fda8a02fde3453656795af5adcd23ef34fa8dbca25d1ab54f9b956344ef51def804128513ba5a3360556c4421388315c9c88429c04d393cc467f38f058f6a0cd7f8051473f20475d78b9db7626bbd234d44b54e3d8289fbb983f30c2b0f88281c11b80e90d3e3ef8a298ea7759eaed772dabf5d189b96e14ce70755d59d58c7efabfe2","0x8000908007d6026e785d1e31ca2641ec57cd629e5552396c9da49d1c9c2980d8bc02176380949e762c4ca1cdb1a5441e5d9c8855ee82bb4fc74673ae55fbefb0cef79c1379"]},"id":1}
```

Reply:
Looks like something is wrong with CI.